Nombre: Abigail Chirinos Espejel  
Clase: 3  
Reto: 03     
Fecha: 11/Agosto/2026  

# Clase 4 - Limpieza de Datos
__DATA Cleaning Challenge__
  
__Objetivo:__ Auditar y limpiar un dataset con problemas reales sin modificar información de manera arbitaria.  
__Challenge:__  
Recibiras un dataset con posibles:  
> * valores faltantes  
> * duplicados  
> * categorias inconsistentes  
> * tipos incorrectos  
> * valores imposibles  
> * outliers  

In [1]:
import pandas as pd 
df_corruptas = pd.read_csv("ventas_challenge.csv")
df_corruptas.head(10)

,ID,Cliente,Producto,Region,Edad,Precio,Fecha,Ventas
0,8714,Alejandro Diaz,Silla,Norte,48,920.36,2025-12-01,1724.72
1,6232,Carlos Ramirez,Monitor,Centro,33,$842.61,Marzo 2 2025,685.67
2,6208,Alejandro Diaz,Monitor,NORTE,30,sin dato,2024-09-11,1413.56
3,1264,Camila Reyes,Teclado,Norte,48,478.63,07-01-2024,1549.10
4,2129,Diego Romero,Mouse,NaN,30,$795.32,2024-12-19,1886.58
5,7536,Valeria Mendoza,LAPTOP,Norte,51,$-454.28,Noviembre 26 2025,1474.70
6,5362,Jose Martinez,Camara,sur,360,1006.99,18/03/2026,1402.18
7,2779,Daniela Cruz,Monitor,Centro,33,$855.42,32/13/2026,1401.90
8,6590,Laura Sanchez,Tablet,Sur,49,907.75,Febrero 22 2026,1316.37
9,9700,Daniela Cruz,Laptop,Sur,42,$659.89,32/13/2026,1280.77


In [2]:
df_corruptas.info()

<class 'pandas.DataFrame'>
RangeIndex: 10300 entries, 0 to 10299
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   ID        10300 non-null  int64  
 1   Cliente   10300 non-null  str    
 2   Producto  10300 non-null  str    
 3   Region    10043 non-null  str    
 4   Edad      10090 non-null  str    
 5   Precio    10300 non-null  str    
 6   Fecha     9987 non-null   str    
 7   Ventas    10300 non-null  float64
dtypes: float64(1), int64(1), str(6)
memory usage: 1.1 MB


In [3]:
df_corruptas.isna().sum()

ID            0
Cliente       0
Producto      0
Region      257
Edad        210
Precio        0
Fecha       313
Ventas        0
dtype: int64

In [4]:
df_corruptas.duplicated().sum()

np.int64(150)

In [5]:
df_corruptas.describe(include='all')

,ID,Cliente,Producto,Region,Edad,Precio,Fecha,Ventas
count,10300.00000,10300,10300,10043,10090,10300,9987,10300.000000
unique,NaN,20,30,22,164,9248,3325,NaN
top,NaN,Carlos Ramirez,Escritorio,Centro,37,sin dato,32/13/2026,NaN
freq,NaN,577,897,1568,338,494,291,NaN
mean,4983.92301,NaN,NaN,NaN,NaN,NaN,NaN,1704.090635
std,2888.22909,NaN,NaN,NaN,NaN,NaN,NaN,5528.343333
min,1.00000,NaN,NaN,NaN,NaN,NaN,NaN,100.000000
25%,2480.75000,NaN,NaN,NaN,NaN,NaN,NaN,1165.745000
50%,4971.50000,NaN,NaN,NaN,NaN,NaN,NaN,1407.515000
75%,7483.25000,NaN,NaN,NaN,NaN,NaN,NaN,1645.080000


| Problema  | Columna | Acción propuesta   |
|:----------|:-------:|--------------------:|
| Faltante  | region  | investigar         |
| Faltante  | edad    | investigar         |
| Faltante  | fecha   | investigar         |
| Duplicado | —       | eliminar si es real|
| Texto     | precio  | normalizar         |
| Texto     | ciudad  | normalizar         |


Primero Analizaremos los duplicados, para despues tratar los faltantes.

In [6]:
#Se crea una tabla con los duplicados para poder analizarlos
duplicados = df_corruptas[df_corruptas.duplicated(keep=False)]
duplicados

,ID,Cliente,Producto,Region,Edad,Precio,Fecha,Ventas
50,404,Andres Ortiz,Escritorio,Norte,38,$-667.47,10/09/2024,1198.38
57,4382,Paola Rios,Laptop,?,49,559.47,Agosto 5 2024,1011.01
74,2730,Miguel Flores,Teclado,Centro,27,505.45,24/01/2024,768.47
88,1749,Pedro Aguilar,Camara,NaN,26,$434.49,2025-08-04,1790.42
92,439,Camila Reyes,Mouse,Norte,42,"$1,586.80",2025-01-11,1823.01
...,...,...,...,...,...,...,...,...
10010,53,Pedro Aguilar,Mouse,Centro,49,$401.01,10/06/2024,1632.07
10188,2756,Alejandro Diaz,Bocina,Occidente,32,1202.38,22/08/2025,1234.78
10219,2422,Diego Romero,Tablet,Norte,35,$757.86,32/13/2026,1947.69
10221,8922,Carlos Ramirez,Impresora,Occidente,33,1041.66,30/08/2024,946.34


In [7]:
#Se ordena la tabla por ID para poder ver los duplicados juntos
duplicados.sort_values(by="ID")

,ID,Cliente,Producto,Region,Edad,Precio,Fecha,Ventas
8682,12,Alejandro Diaz,Mouse,Sur,25,$628.81,2026-05-13,1201.25
360,12,Alejandro Diaz,Mouse,Sur,25,$628.81,2026-05-13,1201.25
328,37,Valeria Mendoza,Escritorio,norte,48,$-640.74,NaN,955.02
7882,37,Valeria Mendoza,Escritorio,norte,48,$-640.74,NaN,955.02
178,53,Pedro Aguilar,Mouse,Centro,49,$401.01,10/06/2024,1632.07
...,...,...,...,...,...,...,...,...
5381,9637,Carlos Ramirez,Monitor,?,37,993.17,05-12-2025,1100.19
8424,9737,Laura Sanchez,Teclado,Sur,44,399.68,2025-01-06,1757.64
1625,9737,Laura Sanchez,Teclado,Sur,44,399.68,2025-01-06,1757.64
1637,9926,Camila Reyes,Monitor,centro,47,$-908.17,2026-05-12,1236.81


In [8]:
df_corruptas["ID"].duplicated().sum()

np.int64(300)

In [9]:
df_corruptas.duplicated().value_counts()

False    10150
True       150
Name: count, dtype: int64

In [10]:
df_corruptas[df_corruptas["ID"].duplicated(keep=False)].sort_values("ID")

,ID,Cliente,Producto,Region,Edad,Precio,Fecha,Ventas
8682,12,Alejandro Diaz,Mouse,Sur,25,$628.81,2026-05-13,1201.25
360,12,Alejandro Diaz,Mouse,Sur,25,$628.81,2026-05-13,1201.25
8098,24,Carlos Ramirez,Silla,Norte,41,"$1,247.73",32/13/2026,2104.40
6960,24,Carlos Ramirez,Silla,Norte,41,"$1,247.73",32/13/2026,2083.64
7882,37,Valeria Mendoza,Escritorio,norte,48,$-640.74,NaN,955.02
...,...,...,...,...,...,...,...,...
1637,9926,Camila Reyes,Monitor,centro,47,$-908.17,2026-05-12,1236.81
4133,9966,Diego Romero,Teclado,Centro,31,1610.60,Febrero 20 2025,1717.42
1072,9966,Diego Romero,Teclado,Centro,31,1610.60,Febrero 20 2025,1725.69
2654,9995,Camila Reyes,ESCRITORIO,Occidente,50,$772.46,Mayo 4 2026,1031.64


In [11]:
df_corruptas[df_corruptas["ID"] == 34]

,ID,Cliente,Producto,Region,Edad,Precio,Fecha,Ventas
6067,34,Pedro Aguilar,Escritorio,Occidente,42,$754.55,2025-04-30,1422.89


### Detección de duplicados

Se identificaron registros duplicados utilizando todas las columnas del
dataset. Para evitar eliminar observaciones válidas, también se analizaron
los ID repetidos.

Se observó que un ID repetido no necesariamente representa un registro
duplicado, ya que algunos registros presentan diferencias en otras variables,
como `Ventas`.

Por esta razón, únicamente se considerarán duplicados aquellos registros
que sean idénticos en todas sus columnas.

In [12]:
# Cantidad de registros duplicados exactos
df_corruptas.duplicated().sum()

np.int64(150)

In [13]:
#Eliminar duplicados exactos, creando un nuevo dataframe
df_sin_duplicados = df_corruptas.drop_duplicates()

In [14]:
#Filas de data set original 
len(df_corruptas)


10300

In [15]:
#Filas de data set sin duplicados
len(df_sin_duplicados)

10150

Cambiar la variable precio, de str a float

In [16]:
# Eliminar el símbolo $ y convertir Precio a float
df_sin_duplicados["Precio"] = pd.to_numeric(
    df_sin_duplicados["Precio"].str.replace("$", "", regex=False),
    errors="coerce")

In [17]:
#Revisar vacios nuevamente
df_sin_duplicados.isna().sum()


ID             0
Cliente        0
Producto       0
Region       253
Edad         207
Precio      1995
Fecha        309
Ventas         0
dtype: int64

Normalizando Fecha

In [18]:
df_sin_duplicados["Fecha"] = pd.to_datetime(
    df_sin_duplicados["Fecha"],
    format="mixed",
    errors="coerce"
)

In [19]:
df_sin_duplicados

,ID,Cliente,Producto,Region,Edad,Precio,Fecha,Ventas
0,8714,Alejandro Diaz,Silla,Norte,48,920.36,2025-12-01,1724.72
1,6232,Carlos Ramirez,Monitor,Centro,33,842.61,NaT,685.67
2,6208,Alejandro Diaz,Monitor,NORTE,30,NaN,2024-09-11,1413.56
3,1264,Camila Reyes,Teclado,Norte,48,478.63,2024-07-01,1549.10
4,2129,Diego Romero,Mouse,NaN,30,795.32,2024-12-19,1886.58
...,...,...,...,...,...,...,...,...
10295,920,Miguel Flores,Camara,Sur,25,954.66,2025-03-28,2084.86
10296,4308,Camila Reyes,Mouse,Sur,38,639.94,2025-05-15,1817.27
10297,5700,Alejandro Diaz,Monitor,Norte,32,619.29,2024-11-03,1304.16
10298,538,Sofia Torres,Bocina,Centro,39,675.27,NaT,1878.64


In [20]:
df_corruptas[df_corruptas["ID"] == 5192]

,ID,Cliente,Producto,Region,Edad,Precio,Fecha,Ventas
3192,5192,Daniela Cruz,Bocina,Sur,24,367.71,09-13-2024,1225.02


In [21]:
# Vuelve a partir del dataset original para no conservar conversiones fallidas
df_limpio = df_corruptas.drop_duplicates().copy()

df_limpio["Precio"] = pd.to_numeric(
    df_limpio["Precio"]
        .astype("string")
        .str.strip()
        .str.replace("$", "", regex=False)
        .str.replace(",", "", regex=False)
        .replace({"sin dato": pd.NA, "": pd.NA}),
    errors="coerce"
)

df_limpio[["Precio"]].isna().sum()

Precio    491
dtype: int64

In [22]:
# Sustituir nombres de meses en español antes de interpretar las fechas
meses = {
    "Enero": "January", "Febrero": "February", "Marzo": "March",
    "Abril": "April", "Mayo": "May", "Junio": "June",
    "Julio": "July", "Agosto": "August", "Septiembre": "September",
    "Octubre": "October", "Noviembre": "November", "Diciembre": "December"
}

fechas = df_limpio["Fecha"].astype("string").str.strip()
for espanol, ingles in meses.items():
    fechas = fechas.str.replace(espanol, ingles, case=False, regex=False)

# mixed admite los distintos formatos presentes en el archivo.
# Revisa las fechas ambiguas (por ejemplo, 05-03-2025) antes de fijar
# si el archivo las expresa como día-mes o mes-día.
df_limpio["Fecha"] = pd.to_datetime(fechas, format="mixed", errors="coerce")

In [23]:
#Revu¿isaremos las variables de Producto y Region
print("PRODUCTO")
print(df_limpio["Producto"].value_counts(dropna=False).to_string())

print("\nREGION")
print(df_limpio["Region"].value_counts(dropna=False).to_string())

PRODUCTO
Producto
Escritorio     878
Tablet         862
Monitor        858
Silla          839
Laptop         825
Mouse          817
Camara         810
Bocina         784
Teclado        780
Impresora      759
LAPTOP         130
TECLADO        109
SILLA          109
 Camara        108
MONITOR        106
 Mouse         104
BOCINA         104
TABLET         103
 Escritorio    102
 Silla          98
MOUSE           97
 Laptop         96
IMPRESORA       93
CAMARA          88
 Impresora      87
ESCRITORIO      84
 Teclado        82
 Monitor        80
 Tablet         80
 Bocina         78

REGION
Region
Centro             1549
Occidente          1490
Sur                1434
Norte              1395
norte               376
occidente           372
centro              365
sur                 364
SUR                 269
OCCIDENTE           262
NaN                 253
NORTE               249
CENTRO              239
 Occidente          196
Sur                 196
Centro              185
 Norte       

Se observa que los datos vienen sucios en producto y region, por lo que como siguiente paso normalizaremos estas variables

In [24]:
# Con este código se eliminan espacios en blanco, se pasa a minúsculas y luego a mayúscula inicial
df_limpio["Producto"] = (
    df_limpio["Producto"]
    .astype("string")
    .str.strip()
    .str.lower()
    .str.title()
)

df_limpio["Region"] = (
    df_limpio["Region"]
    .astype("string")
    .str.strip()
    .str.lower()
    .str.title()
)

print(df_limpio["Producto"].value_counts(dropna=False))
print(df_limpio["Region"].value_counts(dropna=False))

Producto
Escritorio    1064
Laptop        1051
Silla         1046
Tablet        1045
Monitor       1044
Mouse         1018
Camara        1006
Teclado        971
Bocina         966
Impresora      939
Name: count, dtype: int64[pyarrow]
Region
Centro             2488
Occidente          2488
Sur                2419
Norte              2354
<NA>                253
?                    78
Sin Informacion      70
Name: count, dtype: int64[pyarrow]


### Ahora revisaremos faltantes antes de corregir

In [25]:
resumen_faltantes = pd.DataFrame({
    "faltantes": df_limpio.isna().sum(),
    "porcentaje": (df_limpio.isna().mean() * 100).round(2)
})

resumen_faltantes

,faltantes,porcentaje
ID,0,0.00
Cliente,0,0.00
Producto,0,0.00
Region,253,2.49
Edad,207,2.04
Precio,491,4.84
Fecha,595,5.86
Ventas,0,0.00


In [26]:
#Se necesita convertir a numérico para poder hacer análisis estadísticos y visualizaciones. Se eliminan los símbolos de moneda y comas, y se reemplazan los valores "sin dato" y vacíos por NaN.
df_limpio["Precio"] = pd.to_numeric(
    df_limpio["Precio"]
    .astype("string")
    .str.strip()
    .str.replace("$", "", regex=False)
    .str.replace(",", "", regex=False)
    .replace({"sin dato": pd.NA, "": pd.NA}),
    errors="coerce"
)

df_limpio["Edad"] = pd.to_numeric(df_limpio["Edad"], errors="coerce")
df_limpio["Ventas"] = pd.to_numeric(df_limpio["Ventas"], errors="coerce")

print(df_limpio[["Edad", "Precio", "Ventas"]].dtypes)

Edad      float64
Precio    Float64
Ventas    float64
dtype: object


In [27]:
# Convertir a nulos las edades negativas y la edad 397
df_limpio["Edad"] = pd.to_numeric(df_limpio["Edad"], errors="coerce")

edades_invalidas = (df_limpio["Edad"] < 0) | (df_limpio["Edad"] == 397)

print("Edades inválidas encontradas:", edades_invalidas.sum())

df_limpio.loc[edades_invalidas, "Edad"] = pd.NA

print("Nulos en Edad después del cambio:", df_limpio["Edad"].isna().sum())

Edades inválidas encontradas: 130
Nulos en Edad después del cambio: 376


In [28]:
mediana_edad = df_limpio["Edad"].median()
print("Mediana utilizada:", mediana_edad)

df_limpio["Edad"] = df_limpio["Edad"].fillna(mediana_edad)

Mediana utilizada: 38.0


In [29]:
resumen_nulos = pd.DataFrame({
    "Nulos": df_limpio.isna().sum(),
    "Porcentaje": (df_limpio.isna().mean() * 100).round(2)
})

resumen_nulos

,Nulos,Porcentaje
ID,0,0.00
Cliente,0,0.00
Producto,0,0.00
Region,253,2.49
Edad,0,0.00
Precio,491,4.84
Fecha,595,5.86
Ventas,0,0.00


In [31]:
# Uniformar espacios y mayúsculas
df_limpio["Region"] = (
    df_limpio["Region"]
    .astype("string")
    .str.strip()
    .str.title()
)

# Unir las distintas formas de indicar una región desconocida
df_limpio["Region"] = df_limpio["Region"].replace({
    "?": "Sin Especificar",
    "Sin Informacion": "Sin especificar"
})

df_limpio["Region"] = df_limpio["Region"].fillna("Sin Especificar")

df_limpio["Region"].value_counts(dropna=False)

Region
Centro             2488
Occidente          2488
Sur                2419
Norte              2354
Sin Especificar     401
Name: count, dtype: int64[pyarrow]

In [32]:
#Normalizemos Precio
print("Precios nulos:", df_limpio["Precio"].isna().sum())

df_limpio.loc[
    df_limpio["Precio"].isna(),
    ["Producto", "Precio", "Ventas"]
].head(10)

Precios nulos: 491


,Producto,Precio,Ventas
2,Monitor,<NA>,1413.56
20,Teclado,<NA>,1690.45
95,Impresora,<NA>,1885.71
113,Impresora,<NA>,1601.92
148,Mouse,<NA>,745.50
149,Teclado,<NA>,1436.76
164,Camara,<NA>,1344.65
184,Mouse,<NA>,1908.11
199,Bocina,<NA>,714.46
207,Impresora,<NA>,2310.06


In [33]:
mediana_por_producto = df_limpio.groupby("Producto")["Precio"].transform("median")

df_limpio["Precio"] = df_limpio["Precio"].fillna(mediana_por_producto)

print("Precios que siguen nulos:", df_limpio["Precio"].isna().sum())

Precios que siguen nulos: 0


In [34]:
df_limpio.loc[
    df_limpio["Precio"].isna(),
    ["Producto", "Precio", "Ventas"]
]

,Producto,Precio,Ventas


In [35]:
df_limpio.isna().sum()


ID            0
Cliente       0
Producto      0
Region        0
Edad          0
Precio        0
Fecha       595
Ventas        0
dtype: int64

Los precios faltantes se imputaron con la mediana de los precios válidos del mismo producto. Esta decisión conserva las observaciones y toma como referencia productos comparables; los valores imputados son estimaciones, no precios observados.

In [36]:
indices_fecha_nula = df_limpio.index[df_limpio["Fecha"].isna()]

revision_fechas = pd.DataFrame({
    "Fecha_original": df_corruptas.loc[indices_fecha_nula, "Fecha"],
    "Fecha_convertida": df_limpio.loc[indices_fecha_nula, "Fecha"]
})

print("Originalmente vacías:",
      revision_fechas["Fecha_original"].isna().sum())

print("Tenían texto, pero no se convirtieron:",
      revision_fechas["Fecha_original"].notna().sum())

print("\nEjemplos de fechas que no se convirtieron:")
print(
    revision_fechas.loc[
        revision_fechas["Fecha_original"].notna(),
        "Fecha_original"
    ].value_counts().head(20)
)

Originalmente vacías: 309
Tenían texto, pero no se convirtieron: 286

Ejemplos de fechas que no se convirtieron:
Fecha_original
32/13/2026    286
Name: count, dtype: int64


In [37]:
print("Fechas originalmente vacías:",
      revision_fechas["Fecha_original"].isna().sum())

print("Fechas con valor imposible 32/13/2026:",
      revision_fechas["Fecha_original"].eq("32/13/2026").sum())

print("Total de fechas nulas en el dataset limpio:",
      df_limpio["Fecha"].isna().sum())

Fechas originalmente vacías: 309
Fechas con valor imposible 32/13/2026: 286
Total de fechas nulas en el dataset limpio: 595


Después de eliminar los duplicados, se identificaron 309 fechas originalmente vacías y 286 registros con el valor imposible 32/13/2026. Estos últimos se transformaron en valores nulos durante la conversión a fecha. Se conservaron los 595 valores como NaT, ya que no existe información suficiente para recuperar las fechas reales sin inventarlas. Los registros se mantienen porque sus demás variables todavía pueden utilizarse en análisis que no dependan de la fecha

In [38]:
#Ahora revisaremos los valores de Precio y Ventas para ver si hay valores negativos o cero, que no deberían existir en este contexto.
print(df_limpio[["Precio", "Ventas"]].describe())

print("\nPrecios menores o iguales a cero:")
print(df_limpio.loc[df_limpio["Precio"] <= 0,
                    ["ID", "Producto", "Precio", "Ventas"]].head(20))
print("Total:", (df_limpio["Precio"] <= 0).sum())

print("\nVentas menores o iguales a cero:")
print(df_limpio.loc[df_limpio["Ventas"] <= 0,
                    ["ID", "Producto", "Precio", "Ventas"]].head(20))
print("Total:", (df_limpio["Ventas"] <= 0).sum())

           Precio         Ventas
count     10150.0   10150.000000
mean   795.009175    1708.853898
std    412.370015    5568.733852
min      -1768.66     100.000000
25%      634.4225    1166.365000
50%        832.06    1408.020000
75%     1028.7025    1644.885000
max       1903.53  125000.000000

Precios menores o iguales a cero:
       ID    Producto   Precio   Ventas
5    7536      Laptop  -454.28  1474.70
37   2029      Laptop  -509.53  1706.43
50    404  Escritorio  -667.47  1198.38
75   4229      Bocina -1093.01   897.97
76   8512      Laptop  -776.22  1533.26
99   4643     Monitor  -912.68  1681.34
218  1392     Teclado  -907.18  1699.32
220  8167     Monitor -1245.86   634.62
260  2996     Teclado  -1191.2  1097.34
328    37  Escritorio  -640.74   955.02
522  1437     Monitor -1170.59  1686.10
551  5663     Monitor  -999.27  1019.65
586  2828  Escritorio  -848.08   807.78
705  5115      Tablet -1194.54  1929.92
721  9826     Teclado  -773.47  1007.31
735  4797  Escritorio  -855.

In [39]:
for columna in ["Precio", "Ventas"]:
    q1 = df_limpio[columna].quantile(0.25)
    q3 = df_limpio[columna].quantile(0.75)
    iqr = q3 - q1

    limite_inferior = q1 - 1.5 * iqr
    limite_superior = q3 + 1.5 * iqr

    extremos = df_limpio[
        (df_limpio[columna] < limite_inferior) |
        (df_limpio[columna] > limite_superior)
    ]

    print(f"\n{columna}")
    print(f"Límites IQR: {limite_inferior:.2f} a {limite_superior:.2f}")
    print(f"Valores señalados: {len(extremos)}")
    print(extremos[["ID", "Producto", "Precio", "Ventas"]].head(10))


Precio
Límites IQR: 43.00 a 1620.12
Valores señalados: 365
       ID    Producto   Precio   Ventas
5    7536      Laptop  -454.28  1474.70
22    763       Silla  1676.62  1236.92
37   2029      Laptop  -509.53  1706.43
50    404  Escritorio  -667.47  1198.38
75   4229      Bocina -1093.01   897.97
76   8512      Laptop  -776.22  1533.26
99   4643     Monitor  -912.68  1681.34
218  1392     Teclado  -907.18  1699.32
220  8167     Monitor -1245.86   634.62
260  2996     Teclado  -1191.2  1097.34

Ventas
Límites IQR: 448.59 a 2362.66
Valores señalados: 83
        ID   Producto   Precio     Ventas
407   9208     Camara   860.14    2370.75
425   9015     Camara  1119.69     337.91
656   3270     Bocina  1122.21    2429.55
677    563      Mouse   660.49     369.91
824   9764  Impresora  1300.67     418.58
956   4212      Silla   818.92     341.39
1013  5148     Bocina  1050.26     100.00
1350  8631    Teclado   590.88  125000.00
1423  2980      Mouse  1146.16    2475.36
1606  6096      Sill

In [40]:
precios_negativos = df_limpio["Precio"] < 0
precios_cero = df_limpio["Precio"] == 0

print("Precios negativos:", precios_negativos.sum())
print("Precios iguales a cero:", precios_cero.sum())

# Guardar una muestra para documentar qué se corrigió
df_limpio.loc[
    precios_negativos | precios_cero,
    ["ID", "Producto", "Precio", "Ventas"]
].head(10)

Precios negativos: 312
Precios iguales a cero: 0


,ID,Producto,Precio,Ventas
5,7536,Laptop,-454.28,1474.70
37,2029,Laptop,-509.53,1706.43
50,404,Escritorio,-667.47,1198.38
75,4229,Bocina,-1093.01,897.97
76,8512,Laptop,-776.22,1533.26
99,4643,Monitor,-912.68,1681.34
218,1392,Teclado,-907.18,1699.32
220,8167,Monitor,-1245.86,634.62
260,2996,Teclado,-1191.2,1097.34
328,37,Escritorio,-640.74,955.02


In [41]:
# Los precios no positivos se consideran inválidos
df_limpio.loc[df_limpio["Precio"] <= 0, "Precio"] = float("nan")

# Imputar usando únicamente precios válidos del mismo producto
mediana_por_producto = df_limpio.groupby("Producto")["Precio"].transform("median")
df_limpio["Precio"] = df_limpio["Precio"].fillna(mediana_por_producto)

print("Precios nulos restantes:", df_limpio["Precio"].isna().sum())
print("Precios no positivos restantes:", (df_limpio["Precio"] <= 0).sum())

Precios nulos restantes: 0
Precios no positivos restantes: 0


Se detectaron precios negativos, que no son válidos como precios de productos. Se trataron como faltantes y se imputaron con la mediana de precios positivos del mismo producto. Los valores extremos de ventas se señalaron para revisión, sin eliminarlos automáticamente, ya que un importe inusual puede corresponder a una venta real.

In [42]:
print("Ventas más bajas:")
display(
    df_limpio.nsmallest(10, "Ventas")[
        ["ID", "Producto", "Region", "Precio", "Ventas"]
    ]
)

print("Ventas más altas:")
display(
    df_limpio.nlargest(10, "Ventas")[
        ["ID", "Producto", "Region", "Precio", "Ventas"]
    ]
)

print("Ventas menores o iguales a cero:",
      (df_limpio["Ventas"] <= 0).sum())

Ventas más bajas:


,ID,Producto,Region,Precio,Ventas
1013,5148,Bocina,Norte,1050.26,100.00
2333,4641,Impresora,Centro,1114.63,100.00
2863,8420,Monitor,Occidente,1052.67,160.91
7916,2043,Silla,Norte,474.74,232.25
9049,2549,Impresora,Centro,768.6,274.48
3460,6456,Escritorio,Centro,842.39,276.51
4622,5491,Tablet,Sin Especificar,871.59,293.49
6612,903,Bocina,Norte,1275.72,301.01
2938,6289,Laptop,Sur,1178.6,302.36
7185,4414,Tablet,Sur,1271.02,314.35


Ventas más altas:


,ID,Producto,Region,Precio,Ventas
1350,8631,Teclado,Sur,590.88,125000.0
3054,5868,Camara,Norte,922.9,125000.0
4090,6509,Monitor,Norte,838.88,125000.0
4093,2156,Teclado,Sin Especificar,879.24,125000.0
4710,4651,Bocina,Centro,717.56,125000.0
6196,6386,Silla,Centro,641.26,125000.0
6819,7952,Laptop,Sur,529.7,125000.0
7535,4135,Monitor,Sin Especificar,684.36,125000.0
8191,678,Bocina,Norte,611.69,125000.0
2016,4711,Mouse,Sur,595.86,98000.0


Ventas menores o iguales a cero: 0


Se identificaron valores extremos en Ventas mediante el rango intercuartílico. Se revisaron los importes más bajos y altos. Los valores inusuales se conservaron cuando no había evidencia de un error, porque eliminarlos modificaría la distribución y la media de ventas.

In [43]:
print("Filas iniciales:", len(df_corruptas))
print("Filas finales:", len(df_limpio))
print("Duplicados exactos restantes:", df_limpio.duplicated().sum())

print("\nValores faltantes:")
print(df_limpio.isna().sum())

print("\nTipos de datos:")
print(df_limpio.dtypes)

print("\nCategorías de Region:")
print(df_limpio["Region"].value_counts(dropna=False))

print("\nValores imposibles restantes:")
print("Edades negativas o 397:",
      ((df_limpio["Edad"] < 0) | (df_limpio["Edad"] == 397)).sum())
print("Precios no positivos:", (df_limpio["Precio"] <= 0).sum())
print("Ventas no positivas:", (df_limpio["Ventas"] <= 0).sum())

Filas iniciales: 10300
Filas finales: 10150
Duplicados exactos restantes: 0

Valores faltantes:
ID            0
Cliente       0
Producto      0
Region        0
Edad          0
Precio        0
Fecha       595
Ventas        0
dtype: int64

Tipos de datos:
ID                   int64
Cliente                str
Producto            string
Region              string
Edad               float64
Precio             Float64
Fecha       datetime64[us]
Ventas             float64
dtype: object

Categorías de Region:
Region
Centro             2488
Occidente          2488
Sur                2419
Norte              2354
Sin Especificar     401
Name: count, dtype: int64[pyarrow]

Valores imposibles restantes:
Edades negativas o 397: 0
Precios no positivos: 0
Ventas no positivas: 0


In [44]:
df_limpio.to_csv("dataset_limpio.csv", index=False, encoding="utf-8-sig")

print("Archivo guardado: dataset_limpio.csv")
print("Dimensiones:", df_limpio.shape)

Archivo guardado: dataset_limpio.csv
Dimensiones: (10150, 8)
